packages/amplify-graphql-auth-transformer/src/graphql-auth-transformer.ts

```diff
@@ -716,55 +716,133 @@ export class AuthTransformer extends TransformerAuthBase implements TransformerA
+    let redactRelationalField: boolean = needsFieldResolver;
     const relatedModelObject = this.getRelatedModelObject(ctx, getBaseType(field.type));
     const relatedModelName = relatedModelObject.name.value;
     if (this.authModelConfig.has(relatedModelName)) {
       const acm = this.authModelConfig.get(relatedModelName);
-      const roleDefinitions = [
-        ...new Set([
-          ...acm.getRolesPerOperation('get'),
-          ...acm.getRolesPerOperation('list'),
-          ...acm.getRolesPerOperation('sync'),
-          ...acm.getRolesPerOperation('search'),
-          ...acm.getRolesPerOperation('listen'),
-        ]),
-      ].map((r) => this.roleMap.get(r)!);
+      // Get read roles and definitions for related model
+      const relatedModelReadRoleDefinitions = [...new Set(...READ_MODEL_OPERATIONS.map((op) => acm.getRolesPerOperation(op)))].map(
+        (r) => this.roleMap.get(r)!,
+      );
       relatedAuthExpression = this.getVtlGenerator(ctx, relatedModelName).generateAuthExpressionForRelationQuery(
         ctx,
         def,
         field,
         relatedModelObject,
         this.configuredAuthProviders,
-        roleDefinitions,
+        relatedModelReadRoleDefinitions,
         relatedModelObject.fields ?? [],
       );
+      // When the default redaction is false, it means no field resolver is involved
+      // Need the additional check on model level auth rules of both sides to determine the relational field redaction
+      if (!redactRelationalField) {
+        let filteredRelatedModelReadRoleDefinitions = relatedModelReadRoleDefinitions;
+        // When userpool private roles are detected, filter out the non-private userpool roles
+        if (filteredRelatedModelReadRoleDefinitions.some((r) => r.provider === 'userPools' && r.strategy === 'private')) {
+          filteredRelatedModelReadRoleDefinitions = filteredRelatedModelReadRoleDefinitions.filter(
+            (r) => !(r.provider === 'userPools' && r.strategy !== 'private'),
+          );
+        }
+        // When oidc private roles are detected, filter out the non-private oidc roles
+        if (filteredRelatedModelReadRoleDefinitions.some((r) => r.provider === 'oidc' && r.strategy === 'private')) {
+          filteredRelatedModelReadRoleDefinitions = filteredRelatedModelReadRoleDefinitions.filter(
+            (r) => !(r.provider === 'oidc' && r.strategy !== 'private'),
+          );
+        }
+        // relational field read roles are already processed with filter in parent call
+        const fieldReadRoleDefinitions = fieldRoles.map((r) => this.roleMap.get(r)!);
+        /**
+         * Compare if two auth roles are identical
+         * @param role1 auth role 1
+         * @param role2 auth role 2
+         * @returns boolean for two roles are identical
+         */
+        const isIdenticalAuthRole = (role1: RoleDefinition, role2: RoleDefinition): boolean =>
```

Comment:
nit: can you add this as a utliity near the RoleDefinition type itself, and add some unit tests to catch edge cases. Especially to make sure we handle undefined elements in one or both operands.